### PR TITLE
#749: Add StudyStorage.jsm for storing data about studies

### PR DESCRIPTION
**Pull request by Osmose**
This is a continuation of #860. It also relies on #877; commits from 03dc6f9 onward are the interesting ones.

The biggest change between the last PR and this is the use of ajv for schema validation of data. I chose it over other schema validation or rolling our own because we already use JSONSchema elsewhere within Normandy, and I'd rather be consistent in our schemas. One mild downside of this is that types that we could actually store natively in IndexedDB, like dates, are instead stored as strings.

### CODE REVIEW

**Comment by Osmose** [comment]
Rebased to only include the commits that matter, now.

File: recipe-client-addon/lib/StudyStorage.jsm [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    const db = await getDatabase();
+    return getStore(db).add(study);
+  },
+
+  async update(studyName, data) {
+    const db = await getDatabase();
+    const savedStudy = await getStore(db).get(studyName);
+    if (!savedStudy) {
+      throw new Error(`Cannot update study ${studyName}: could not find study.`);
+    } else if (!validateUpdate(data)) {
```
Comment (gijsk) [inline recipe-client-addon/lib/StudyStorage.jsm]: Nit: no else after return/throw.

File: recipe-client-addon/lib/StudyStorage.jsm [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+const DB_OPTIONS = {
+  version: 1,
+  storage: "persistent",
+};
+const DB_SCHEMA = {
+  type: "object",
+  additionalProperties: false,
+  properties: {
+    name: {
+      type: "string",
```
Comment (gijsk) [inline recipe-client-addon/lib/StudyStorage.jsm]: Based on a *very* quick read of docs / the interwebz (I tried to find a live demo site but couldn't), I *think* this doesn't require `name` or any of the other string props to be non-empty, so passing `{name: "", addonId: "", addonVersion: "", description: ""}` would be valid. Should we fix that (seems like `minLength: 1` on those props might be enough, though we could try to do regex validation of the versions) ?

**Comment by Osmose** [comment]
Feedback'd.

**Comment by mythmon** [comment]
Can we leave this PR open and unmerged for a bit? I've got a reply from GitHub support about our CI issue, and having an open PR that shows the issue should help.

**Comment by mythmon** [comment]
Closing and re-opening to help debug CI not running on PRs.

File: recipe-client-addon/lib/StudyStorage.jsm [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  async clear() {
+    const db = await getDatabase();
+    await getStore(db).clear();
+  },
+
+  async close() {
+    if (databasePromise) {
+      const db = await databasePromise;
+      await db.close();
+      databasePromise = null;
```
Comment (gijsk) [inline recipe-client-addon/lib/StudyStorage.jsm]: Seconded! I noted this in https://github.com/mozilla/normandy/pull/860#discussion_r126279145 and then forgot about it when reviewing this version of the PR. Oops...

**Comment by Osmose** [comment]
Feedback'd. Any chance we can create a new PR to demonstrate this issue instead of blocking this? It's starting to block other work.

**Comment by mythmon** [comment]
I think that we've got everything we can out of this for now. Lets go ahead and merge it.

**Comment by mythmon** [comment]
Using admin powers to merge this PR since CI isn't running.

**Pull request merged**